### ISSUE TITLE
NullReferenceException when using either --inprocess or --domain=Multiple options

### ISSUE THREAD

**Issue by kirillonufreinko** [nunit/nunit#1732]
Due to a problem in project set-up (it appears so) I receive NullReferenceException when running unit test with following stack trace:
```
Test Run Summary
  Overall result: System.NullReferenceException: Object reference not set to an instance of an object.
   at NUnit.Common.ColorConsoleWriter.WriteLabel(String label, Object option, ColorStyle valueStyle)
  at NUnit.Common.ColorConsoleWriter.WriteLabelLine(String label, Object option, ColorStyle valueStyle)
 at NUnit.ConsoleRunner.ResultReporter.WriteSummaryReport()
 at NUnit.ConsoleRunner.ConsoleRunner.RunTests(TestPackage package, TestFilter filter)
  at NUnit.ConsoleRunner.Program.Main(String[] args)
```
No other evidence of the problem are there in report or console.
When attached with VS to the nunit process I've managed to discover that the report is the following:
```
<test-run id="2" testcasecount="0" engine-version="3.4.1.0" clr-version="4.0.30319.42000" start-time="2016-08-10 12:48:48Z" end-time="2016-08-10 12:48:48Z" duration="0.049014">
  <command-line><![CDATA["F:\Projects\[censored]\./packages/NUnit.ConsoleRunner.3.4.1/tools/nunit3-console.exe" F:\Projects\[censored]\UnitTests.nunit]]></command-line>
 <test-suite type="Project" id="2" name="UnitTests.nunit" fullname="F:\Projects\[censored]\UnitTests.nunit" testcasecount="0" />
</test-run>
```
It appears NullReferenceException happens because ResultReporter relies on the fact that "result" property exists in test run which is not true:
```
public ResultReporter(XmlNode result, ExtendedTextWriter writer, ConsoleOptions options)
{
    _result = result;
    _writer = writer;

    _overallResult = result.GetAttribute("result");
    if (_overallResult == "Skipped")
        _overallResult = "Warning";

    _options = options;

    Summary = new ResultSummary(result);
}
```

Could we improve logging in this case?

**Comment by ChrisMaddock** [comment]
Hi @kirillonufreinko - can you clarify what the issue was with your project set-up? I presume you're referring to an error in UnitTests.nunit?

**Comment by ChrisMaddock** [comment]
@kirillonufreinko - are you able to clarify exactly what the problem in your project set-up was that produced that xml?

**Comment by kirillonufreinko** [comment]
I'm trying to figure out this right now. At the current moment it seems like if I have more then one test assembly in then project - it fails. Probably it's wrong - I did not finish yet

**Comment by kirillonufreinko** [comment]
So I've tried several times and several different set-ups,edited file manually and with nunit project editor. The nullreference happens when I add more the one assembly to test project

**Comment by ChrisMaddock** [comment]
Can you post a project file that causes the error for you? I'm unable to replicate it at the moment.

**Comment by kirillonufreinko** [comment]
Here is what nunit project is (but for the exact names of assemblies)

<NUnitProject>
  <Settings activeconfig="Debug" domainUsage="Multiple" processModel="Default" />
  <Config name="Debug" binpathtype="Auto">
    <assembly path="Bin\Debug\[censored1].dll" />
    <assembly path="Bin\Debug\[censored2].dll" />
    <assembly path="Bin\Debug\[censored3].dll" />                                         
    <assembly path="Bin\Debug\[censored4].dll" />                                  
    <assembly path="Bin\Debug\[censored5].dll" />                               

  </Config>
</NUnitProject>

**Comment by ChrisMaddock** [comment]
Thanks for that! Does it work if you switch `domainUsage="Multiple"` to `domainUsage="Default"`?

**Comment by kirillonufreinko** [comment]
It works with domainUsage="Default"

**Comment by ChrisMaddock** [comment]
I think this is a settings collision. 

You posted:
`<Settings activeconfig="Debug" domainUsage="Multiple" processModel="Default" />`

With multiple assemblies - `processModel=Default` resolves to running each assembly in it's own process. `domain-usage=multiple` is asking nunit run each assembly in each process in a separate assembly - which I presume is falling over when it finds only one assembly per process. It's not an area I'm familiar with however - maybe someone will correct me later!

In the meantime, try:

`<Settings activeconfig="Debug" domainUsage="Default" processModel="Default" />`

This should run each assembly in it's own process, and thus appdomain.

**Comment by ChrisMaddock** [comment]
For repro, here's a project file which causes the exception:
```
<NUnitProject>
  <Settings activeconfig="Debug" domainUsage="Multiple" processModel="Default" />
  <Config name="Debug" binpathtype="Auto">
    <assembly path="Bin\Debug\[censored1].dll" />
    <assembly path="Bin\Debug\[censored2].dll" />
  </Config>
</NUnitProject>
```

**Comment by CharliePoole** [comment]
@ChrisMaddock I'll have to give this a shot. Does the repro work with any named assemblies or do they actually have to exist?

The settings are not actually contradictory in their intended interpretation. The intent is that DomainUsage.Multiple will use one domain per assembly run in the process, which would be 1 in this case. But it's possible that some code change has introduced the bug.

**Comment by ChrisMaddock** [comment]
@CharliePoole - yeah, the above project file works just as is, it doesn't get as far as looking for the files, I guess. I also wanted to try it with the equivalent command line options to see if the project file was even relevant - but ran out of lunchbreak!

**Comment by CharliePoole** [comment]
Is this the NUnit 3.4.1 release, btw?

As it turns out, I'm seeing a different problem in the latest master.

**Comment by kirillonufreinko** [comment]
It's 3.4.1

**Comment by ChrisMaddock** [comment]
> I also wanted to try it with the equivalent command line options to see if the project file was even relevant - but ran out of lunchbreak!

@CharliePoole - I tried it, it's not. The below command line also causes the exception:

`nunit3-console a.dll b.dll --process=Multiple --domain=Multiple`

**Comment by CharliePoole** [comment]
I can confirm this simpler repro in the current code as well. The repro using a project is hidden because the current code seems to have forgotten how to recognize projects!

**Comment by CharliePoole** [comment]
Also errors if you do

```
nunit3-console a.dll --inprocess --domain-Multiple
```

So it's really not the combination but the multiple domain option by itself.

**Comment by CharliePoole** [comment]
And finally, the same error appears if you use

```
nunit3-console a.dll --inprocess
```

So, in summary, this error appears whenever you use --inprocess or --domain:Multiple. It appears to have crept in with the fixes I made to 3.4.1. Unfortunately, I think this qualifies as critical. What do you think @rprouse?

**Comment by rprouse** [comment]
I agree it is critical. We should probably consider a 3.4.2.

**Comment by rprouse** [comment]
@CharliePoole I tried cherry picking #1736 into 3.4.1 to create a 3.4.2 hotfix, but most of those changes are fixes to Minimize use of agents in loading tests for execution #1718. I am going to test the various combinations and determine if these issues exist in 3.4.1 or if some of them were introduced by #1718. I will then make a call as to if it warrants a hotfix.

**Comment by rprouse** [comment]
I used the following command lines with two test assemblies using 3.4.1.

## Works

- [x] `--inprocess` and `--process=InProcess`
- [x] `--process=Separate`
- [x] `--process=Multiple`
- [x] `--domain=Single`
- [x] `--domain=Multiple`
- [x] `--inprocess --domain=Single`
- [x] `--process=Separate --domain=Single`
- [x] `--process=Multiple --domain=Single`
- [x] `--inprocess --domain=Multiple`

## Errors

`--domain=None` or `--inprocess --domain=None` fails with `Could not load file or assembly 'nunit.framework' or one of its dependencies. The system cannot find the file specified.` I am running using the installed console from the test project bin directory.

`--domain=Multiple` fails with,

```
  Overall result: System.NullReferenceException: Object reference not set to an instance of an object.
   at NUnit.Common.ColorConsoleWriter.WriteLabel(String label, Object option, ColorStyle valueStyle)
   at NUnit.Common.ColorConsoleWriter.WriteLabelLine(String label, Object option, ColorStyle valueStyle)
   at NUnit.ConsoleRunner.ResultReporter.WriteSummaryReport()
   at NUnit.ConsoleRunner.ConsoleRunner.RunTests(TestPackage package, TestFilter filter)
   at NUnit.ConsoleRunner.Program.Main(String[] args)
```

## Conclusions

1. We have an issue with `--domain=None` not finding referenced assemblies in the bin directory. #1741
2. `--domain=Multiple` fails running in the agent (separate process). Odd that causes an issue with `WriteLabel` though and that it works in-process. #1742

I have created issues for both of these, but I don't think this is serious enough for a hotfix. Anyone disagree?

**Comment by CharliePoole** [comment]
@rprouse It sounds to me as if something went wrong with the cherry-pick. If you look at #1736 itself, it has minimal changes.

The sequence of tests you ran are pretty much what I did with exceptions:
 * I didn't do `domain=none` for reasons explained in my comment to #1741 
 * I didn't separately test `process=Multiple` versus no `process` option, since Multiple is the default
 * I ran with a single assembly - this may be the key difference.

Once we get this in (or before?) we should automate something like this for pre-release testing.
